keps/sig-node/3759-kep-template/README.md

```diff
@@ -0,0 +1,1209 @@
+
+### Problems: logging/metrics sidecar
+
+A logging sidecar should start before several other containers, to not lose logs
+from the startup of other applications. Let's call "main" container the app that
+will log and "logging" container the sidecar that will facilitate it.
+
+If the logging container starts after the main app, some logs can be lost.
+Furthermore, if the logging container is not yet started and the main app
+crashes on startup, those logs can be lost (depends if logs go to a shared
+volume or over the network on localhost, etc.). While you can modify your
+application to handle this scenario during startup (as it is probably the change
+you need to do to handle sidecar crashes), for shutdown this approach won't
+work.
+
+On shutdown the ordering behavior is arguably more important: if the logging
+container is stopped first, logs for other containers are lost. No matter if
+those containers queue them and retry to send them to the logging container, or
+if they are persisted to a shared volume. The logging container is already
+killed and will not be restarted, as the pod is shutting down. In these cases,
+logs are lost.
+
+The same things regarding startup and shutdown apply for a metrics container.
+
+### Problems: service mesh
+
+Service mesh presents a similar problem: you want the service mesh container to
+be running and ready before other containers start, so that any inbound/outbound
+connections that a container can initiate goes through the service mesh.
+
+A similar problem happens for shutdown: if the service mesh container is
+terminated prior to the other containers, outgoing traffic from other apps will
+be blackholed or not use the service mesh.
+
+However, as none of these are possible to guarantee, most service meshes (like
+Linkerd and Istio), need to do several hacks to have the basic functionality.
+
+Some service meshes depend on secret (like a certificate) downloaded by other
+Init container to establish secure communication between services. This makes
+the problem of ordering of sidecar and Init containers harder.
+
+### Problems: configuration / secrets
+
+Some pods use init containers to pull down configuration/secrets and update them
+before the main container gets it. Then use sidecars to continue to watch for
+changes and perform the updates and push to the main container. This requires
+two separate code paths today. Perhaps the same sidecar could be used for both
+cases.
+
+### Goals
+
+This proposal aims to:
+- make containers implementing the sidecar pattern first class citizens inside a
+  Pod
+- solve a Job completion issue when sidecars should run continuously
+- allow mixing initContainers and sidecars for a choreographed startup sequence
+- allow easy injection of sidecar containers in any Pod
+
+### Non-Goals
+
+This proposal doesn't aim to:
+- support arbitrary dependencies graphs between containers
+- (alpha) support containers termination ordering
+
```

Comment:
One scenario that comes to mind is where a user assumes that this feature will ensure that their workloads don't run without their security related sidecar container also running.  If the sidecar starts successfully and becomes ready, but then fails repeatedly, that won't be the case.

This proposal doesn't aim to:
- act as a security control to enforce that pod containers only run while the sidecar is healthy

Reply:
added. Good call